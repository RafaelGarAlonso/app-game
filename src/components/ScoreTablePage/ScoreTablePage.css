.score-table-page-title {
    text-align: center;
    font-weight: bold;
    font-size: 30px;
    margin: 10px 0 20px 0;
}

.table {
    overflow-y: auto;
    max-height: 480px;
    overflow-x: hidden;
}