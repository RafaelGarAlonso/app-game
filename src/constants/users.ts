export const INIT_USER = {
    "userName": "",
    "points": 0
};

export const INIT_USERS_TABLE = [
    {
        userName: "Manolo",
        points: 100
    },
    {
        userName: "Pepe",
        points: 88
    },
    {
        userName: "Benito",
        points: 50
    }
];

export const listOfPlays = ["rock", "paper", "scissors"];