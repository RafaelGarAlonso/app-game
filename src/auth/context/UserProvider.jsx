import { UserContext } from './UserContext';
import { INIT_USER, INIT_USERS_TABLE } from '../../constants/users'

export const UserProvider = ({ children }) => {

  const getUser = () => {
    const localStorageData = localStorage.getItem('currentUser');
    return localStorageData ? JSON.parse(localStorageData) : INIT_USER;
  };

  const setNewUserSession = ({ userName }) => {
    const listOfUsers = getListOfUsers();

    for (const [index, user] of listOfUsers.entries()) {
      if (user.userName === userName) {
        const { userName, points } = listOfUsers[index];
        localStorage.setItem('currentUser', JSON.stringify({ userName, points }));
        return false;
      }
    };

    localStorage.setItem('currentUser', JSON.stringify({ userName, points: 0 }));
    return true;
  };

  const getAppLoaded = () => {
    return sessionStorage.getItem('appLoaded');
  };

  const setAppLoaded = (loaded) => {
    sessionStorage.setItem('appLoaded', loaded);
  };

  const userEarnPoints = () => {
    const user = getUser();
    localStorage.setItem('currentUser', JSON.stringify({ userName: user.userName, points: user.points + 1 }));
  };

  const getListOfUsers = () => {
    const listOfUsers = localStorage.getItem('listOfUsers');
    return listOfUsers ? JSON.parse(listOfUsers) : null;
  };

  const setNewListOfUsers = (listOfUsers = INIT_USERS_TABLE) => {
    localStorage.setItem('listOfUsers', JSON.stringify(listOfUsers));
  };

  const setPointsToListUsers = (currentUser) => {
    const listOfUsers = getListOfUsers();

    for (let [index, user] of listOfUsers.entries()) {
      if (user.userName === currentUser.userName) {
        listOfUsers[index].points = currentUser.points;
        break;
      }
    };

    setNewListOfUsers(listOfUsers);
  };

  return (
    <UserContext.Provider value={{
      setNewUserSession,
      getUser,
      setAppLoaded,
      getAppLoaded,
      userEarnPoints,
      getListOfUsers,
      setNewListOfUsers,
      setPointsToListUsers
    }}>
      {children}
    </UserContext.Provider>
  )
}