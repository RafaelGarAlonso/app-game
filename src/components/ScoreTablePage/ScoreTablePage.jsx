
import { useNavigate } from 'react-router-dom';
import { Button } from 'react-bootstrap';
import { DataTable } from '../DataTable/DataTable';
import { useContext } from 'react';
import { UserContext } from '../../auth/context/UserContext';
import './ScoreTablePage.css'

const DATA_TABLE_HEADERS = ['#', 'Username', 'Points'];

export const ScoreTablePage = () => {
    const navigate = useNavigate();
    const { getListOfUsers } = useContext(UserContext);

    const onBack = () => {
        navigate('/home');
    };

    const loadRows = () => {
        return getListOfUsers() ? getListOfUsers().sort((a, b) => b.points - a.points) : [];
    }

    return (
        <>
            <h1 className="score-table-page-title">Hall of Fame</h1>
            <div className="table">
                <DataTable headers={DATA_TABLE_HEADERS} rows={loadRows()} />
            </div>
            <Button onClick={onBack} className="button-back" size="lg" variant="dark">Exit</Button>
        </>
    )
}
