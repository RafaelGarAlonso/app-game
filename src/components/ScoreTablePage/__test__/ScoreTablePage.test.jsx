import { render, fireEvent } from '@testing-library/react';
import { ScoreTablePage } from '../ScoreTablePage';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../../auth/context/UserProvider';
import { INIT_USERS_TABLE } from '../../../constants/users'

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom')),
    useNavigate: () => mockedUsedNavigate,
}));

describe('Test <ScoreTablePage />', () => {

    const renderTemplate = (INIT_USERS_TABLE) => {
        return render(
            <UserProvider getListOfUsers={INIT_USERS_TABLE}>
                <MemoryRouter>
                    <ScoreTablePage />
                </MemoryRouter>
            </UserProvider>
        );
    }

    test('Check snapshot component', () => {
        const { container } = renderTemplate(INIT_USERS_TABLE);
        expect(container).toMatchSnapshot();
    });


    test('It should click on button back and navigate to /home', () => {
        const { container } = renderTemplate();
        const backButton = container.querySelector(".button-back");

        fireEvent.click(backButton);
        expect(mockedUsedNavigate).toBeCalled();
    });

});