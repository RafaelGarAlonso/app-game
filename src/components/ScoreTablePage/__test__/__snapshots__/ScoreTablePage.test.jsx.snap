// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <ScoreTablePage /> Check snapshot component 1`] = `
<div>
  <h1
    class="score-table-page-title"
  >
    Hall of Fame
  </h1>
  <div
    class="table"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <table
          class="table table table-sm table-striped table-bordered"
        >
          <thead>
            <tr>
              <th>
                #
              </th>
              <th>
                Username
              </th>
              <th>
                Points
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                EMPTY LIST
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    class="button-back btn btn-dark btn-lg"
    type="button"
  >
    Exit
  </button>
</div>
`;
